### PR TITLE
#100 display multiple messages on token

### LINKED ISSUE #100: Handle and display multiple messages on Tokens
There may be more than one error associated with a Token, or a suggestion along with an error. We should change the message display from a tooltip to a popup, similar to the lookup results.

This paves the way for #101 , where we can auto-fix certain errors/suggestions.

### PR DESCRIPTION
**Pull request by craigp-atw**
- Added support for more message types (info and warning)
- Added support to show multiple messages (resolves #100 )

image
image
image

- Fixed an issue where a word was being flagged as 'ambiguous sense' even when a sense was specified.
image

### CODE REVIEW

File: app/src/index.d.ts [hunk 1]
```diff
@@ -25,12 +24,20 @@ type Sentence = {
...
-	error?: string
-	suggest?: string
+	message_type: MessageType
+	message: string
+	// TODO #101 add optional fix_action structure
 }
 
-interface MessageInfo extends Message {
+interface MessageInfo {
+	// TODO find a more elegant way to do this?
```
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/index.d.ts:36]: but maybe you were looking for something like this?

```js
type MessageType = 'error' | 'warning' | 'suggest' | 'info';

interface Message {
  [key in MessageType]: string;
}
```

File: app/src/lib/rules/case_frame/common.js [hunk 2]
```diff
@@ -343,11 +343,17 @@ export function* validate_case_frame(trigger_context) {
 	const selected_result = token.lookup_results[0]
 	const case_frame = selected_result.case_frame
 
+	if (!case_frame.is_valid) {
+		yield { error: 'Invalid argument structure for {sense}. Consult the Ontology for correct usage.' }
```
Comment (longrunningprocess) [inline app/src/lib/rules/case_frame/common.js:347]: I did not realize you were using generator functions... I've never used them, but I'd love to learn more about their utility for what you're doing.

File: app/src/lib/test_helps.js
Comment (longrunningprocess) [inline app/src/lib/test_helps.js:1]: let's make sure this isn't getting bundled... there's probably a config to ensure it's excluded.

File: app/src/lib/tokens/Message.svelte [hunk 3]
```diff
@@ -1,23 +1,35 @@
...
+	import PopupMenu from './PopupMenu.svelte'
+	import Table from './Table.svelte'
 	import { Badge } from '$lib'
 	import Icon from '@iconify/svelte'
+	import { token_has_message } from '$lib/parser/token'
 
 	/** @type {SimpleToken} */
 	export let token
+
+	const message_ui = {
```
Comment (longrunningprocess) [inline app/src/lib/tokens/Message.svelte:11]: slick, I like this in place of an `if`, well done!

File: app/src/routes/check/+server.js [hunk 4]
```diff
@@ -103,4 +102,22 @@ function simplify_tokens(sentences) {
 			missing_arguments: missing_arguments.map(rule => rule.role_tag),
 		}
 	}
+}
+
+/** @type {Map<MessageType, number>} */
+const MESSAGE_SEVERITY = new Map([
```
Comment (longrunningprocess) [inline app/src/routes/check/+server.js:108]: I wonder about putting `severity` on the `Message` and letting the error constructor handle these assignments...

File: app/src/lib/tokens/PopupMenu.svelte [hunk 5]
```diff
@@ -1,6 +1,9 @@
+<script>
+	export let color_classes = 'bg-info text-info-content'
```
Comment (longrunningprocess) [inline app/src/lib/tokens/PopupMenu.svelte:2]: I'm wondering if this is premature... I only see it being used in one place at this time.

Comment (craigp-atw) [reply app/src/lib/tokens/PopupMenu.svelte:2]: `Result` and `NotFound` use `PopupMenu` and use the default info colors. The new `Message` needed to use different colors. I'm open to other suggestions though!

File: app/src/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply app/src/index.d.ts:36]: we discussed a "const" approach while pairing.

File: app/src/lib/rules/case_frame/common.js (hunk 2 above)
Comment (longrunningprocess) [reply app/src/lib/rules/case_frame/common.js:347]: explained while pairing.

File: app/src/lib/test_helps.js
Comment (longrunningprocess) [reply app/src/lib/test_helps.js:1]: further research being done on this approach.

File: app/src/routes/check/+server.js (hunk 4 above)
Comment (longrunningprocess) [reply app/src/routes/check/+server.js:108]: this will be incorporated into next commits.

**Review by longrunningprocess — APPROVED** [review]
This is much more reassuring, thank you.

**Pull request merged**